velox/functions/prestosql/types/IPAddressType.cpp

```diff
@@ -0,0 +1,238 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/prestosql/types/IPAddressType.h"
+#include <folly/IPAddress.h>
+#include "velox/expression/CastExpr.h"
+
+static constexpr int kIPV4AddressBytes = 4;
+static constexpr int kIPV4ToV6FFIndex = 10;
+static constexpr int kIPV4ToV6Index = 12;
+static constexpr int kIPAddressBytes = 16;
+static constexpr int kIPAddressMaxStrLen = 39;
+
+namespace facebook::velox {
+
+namespace {
+
+class IPAddressCastOperator : public exec::CastOperator {
+ public:
+  bool isSupportedFromType(const TypePtr& other) const override {
+    switch (other->kind()) {
+      case TypeKind::VARBINARY:
+      case TypeKind::VARCHAR:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  bool isSupportedToType(const TypePtr& other) const override {
+    switch (other->kind()) {
+      case TypeKind::VARBINARY:
+      case TypeKind::VARCHAR:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  void castTo(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, resultType, result);
+
+    if (input.typeKind() == TypeKind::VARCHAR) {
+      castFromString(input, context, rows, *result);
+    } else if (input.typeKind() == TypeKind::VARBINARY) {
+      castFromVarbinary(input, context, rows, *result);
+    } else {
+      VELOX_UNSUPPORTED(
+          "Cast from {} to IPAddress not supported", resultType->toString());
+    }
+  }
+
+  void castFrom(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, resultType, result);
+
+    if (resultType->kind() == TypeKind::VARCHAR) {
+      castToString(input, context, rows, *result);
+    } else if (resultType->kind() == TypeKind::VARBINARY) {
+      castToVarbinary(input, context, rows, *result);
+    } else {
+      VELOX_UNSUPPORTED(
+          "Cast from IPAddress to {} not supported", resultType->toString());
+    }
+  }
+
+ private:
+  static void castToString(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      BaseVector& result) {
+    auto* flatResult = result.as<FlatVector<StringView>>();
+    const auto* ipaddresses = input.as<SimpleVector<int128_t>>();
+    folly::ByteArray16 addrBytes;
+
+    flatResult->resize(rows.size());
+    flatResult->getRawStringBufferWithSpace(rows.size() * kIPAddressMaxStrLen);
```

Comment:
you dont need this, since you are using  `exec::StringWriter`, it takes care of string buffer management for you.

In fact calling getRawStringBufferWithSpace() would be counter productive as this would try to allocate a single large buffer of this size, whereas the vector might already have a number of smaller buffers that can accomodate values that you will be writing next. Using `exec::StringWriter` will abstract away all those lower level details for you so would be the preferred way to go.

Another way to do this (without using `exec::StringWriter`) would be to use the FlatVector's API `FlatVector<StringView>::set(vector_size_t idx, StringView value)` which also takes care of buffer management

Reply:
@bikramSingh91 Thank you for the tips. The reason I tried to use the allocator was via @mbasmanova's suggestion of preallocating the strings before hand. See
https://github.com/facebookincubator/velox/pull/10596#discussion_r1700872255
&
https://github.com/facebookincubator/velox/pull/10596#discussion_r1710433226

At first I didn't have any reservation for the string space, then I reserved each individual row's string space. But then @mbasmanova suggested I allocate the entire string space before entering the row by row casting.